### PR TITLE
*: optimize lock usage

### PR DESCRIPTION
**Pull request by rleungx**
Signed-off-by: Ryan Leung <rleungx@gmail.com>

### What problem does this PR solve?

Issue Number: Ref #4515.

### What is changed and how it works?

This PR removes some necessary lock.

```commit-message
```

### Check List

Tests

- Unit test

### Release note

```release-note
None
```

### CODE REVIEW

**Review by HunDunDM — COMMENTED** [review]
I'm worried about some nil pointers after breaking consistency.

File: server/cluster/cluster.go [hunk 1]
```diff
@@ -217,10 +212,7 @@ func (c *RaftCluster) InitCluster(
 	c.ctx, c.cancel = context.WithCancel(c.serverCtx)
```
[the remark this answers is not in the document]
Comment (rleungx) [reply server/cluster/cluster.go]: I checked the RPC calls and only `SplitRegions` that needs to be modified.

File: server/cluster/cluster.go [hunk 2]
```diff
@@ -1136,9 +1155,6 @@ func (c *RaftCluster) UpStore(storeID uint64) error {
 
 // SetStoreWeight sets up a store's leader/region balance weight.
 func (c *RaftCluster) SetStoreWeight(storeID uint64, leaderWeight, regionWeight float64) error {
-	c.Lock()
```
Comment (bufferflies) [inline server/cluster/cluster.go:1139]: Is it safe if update store weight by multi thread?

**Review by disksing — APPROVED** [review]

File: server/cluster/cluster.go [hunk 3]
```diff
@@ -418,62 +413,106 @@ func (c *RaftCluster) Context() context.Context {
...
 }
 
-// GetHeartbeatStreams returns the heartbeat streams.
-func (c *RaftCluster) GetHeartbeatStreams() *hbstream.HeartbeatStreams {
-	c.RLock()
-	defer c.RUnlock()
-	return c.coordinator.hbStreams
+// GetMergeChecker returns merge checker.
+func (c *RaftCluster) GetMergeChecker() *checker.MergeChecker {
+	return c.coordinator.checkers.GetMergeChecker()
```
Comment (AndreMouche) [inline server/cluster/cluster.go:438]: Removing the `Lock` LGTM, however, should we consider the situation that `c.coordinator` is nil(IsRunning==false)?  
Here contains two issues:
1. panic with nil: if the c.coordinator is nil
2. data race: if there exists a goroutine that is starting the server(updating the c.coordinator) at the same time

If we are the only programmer who knows how to use the lib, we should run `c.Start` successfully at first. However the later programmer may not know this condition, so to be a better `lib`, Could you please check `c.IsRunning` first(which can both avoid the nil issue and keep thread safe)?

**Review by AndreMouche — COMMENTED** [review]
Rest LGTM

File: server/cluster/cluster.go [hunk 4]
```diff
@@ -1430,31 +1441,23 @@ func (c *RaftCluster) resetMetrics() {
 }
 
 func (c *RaftCluster) collectClusterMetrics() {
-	c.RLock()
 	if c.regionStats == nil {
```
[the remark this answers is not in the document]
Comment (rleungx) [reply server/cluster/cluster.go:1432]: This is guaranteed by our internal logic.

File: server/statistics/region_collection.go [hunk 5]
```diff
@@ -274,6 +287,8 @@ func NewLabelStatistics() *LabelStatistics {
 
 // Observe records the current label status.
 func (l *LabelStatistics) Observe(region *core.RegionInfo, stores []*core.StoreInfo, labels []string) {
+	l.Lock()
+	defer l.Unlock()
```
Comment (AndreMouche) [inline server/statistics/region_collection.go:291]: The lock could start after L293

File: server/statistics/region_collection.go [hunk 6]
```diff
@@ -308,6 +327,8 @@ func (l *LabelStatistics) ClearDefunctRegion(regionID uint64) {
 
 // GetLabelCounter is only used for tests.
 func (l *LabelStatistics) GetLabelCounter() map[string]int {
+	l.RLock()
```
Comment (AndreMouche) [inline server/statistics/region_collection.go:330]: Although it is only used in tests,  there are two issues:
- First, the lock is unnecessary
- secondary, It is not threadsafe to return the pointer(l.labelCounter)

**Review by AndreMouche — APPROVED** [review]
LGTM

**Pull request merged**